src/cluster.h

```diff
@@ -173,257 +51,102 @@ typedef struct {
     redisDb *db;                /* A link back to the db this dict belongs to */
 } clusterDictMetadata;
 
+/*-----------------------------------------------------------------------------
+ * Cluster public structs
+ *----------------------------------------------------------------------------*/
+
+typedef struct clusterNode {
+    char name[CLUSTER_NAMELEN]; /* Node name, hex string, sha1-size */
+    char shard_id[CLUSTER_NAMELEN]; /* shard id, hex string, sha1-size */
+    void* data;
+} clusterNode;
+
 typedef struct clusterState {
```

Comment:
I want to think through the benefit of keeping the slot->node and shard->node mapping here, the "Redis topology state". It seems like it might allow a lot more of shared code between the implementations, but at the same time it might make the interface much harder since now the clustering implementations need to have a complex update operation.

Reply:
Yes. Totally. Also undecided about this.

Reply:
I spent like an hour prototyping it, and my conclusion is probably this is fine. Basically:
**Pros:***
- Cluster shards, slots and some other misc commands become much cleaner, since we have all the state to show them.
- The API contract is smaller by a couple of APIs. We can compact a lot of stuff together in the updates, not sure if that is a huge win.

**Cons:**
- We basically need inverse APIs for the node update operations, which are arguably more annoying to orchestrate since you have to understand when to call them. 
- Each implementation still needs it's own map of data, so there is quite a bit of duplication. The data doesn't need to be same I suppose. Duplication of data is worse than duplication of code IMO, lower chance of data drift.

Reply:
"CLUSTER SHARDS" is already in cluster.c and works with the current exposed API. "CLUSTER SLOTS" on the other hand is very enmeshed in the internal "legacy" implementation, partially for optimization. In the mean time I've left it where it is assuming that we will either create a single more generic implementation later or flotilla will implement it for itself from scratch.

Reply:
Also, the main parts of the cluster implementation, the parts that serve data, are currently generic implementations in cluster.c. Would it be nice to have more shared code? Definitely, agree with your cost/benefit analysis that it's not worth it in this case.